pegen/c_generator.py

```diff
@@ -145,6 +146,31 @@ class CParserGenerator(ParserGenerator, GrammarVisitor):
     def __init__(self, rules: Dict[str, grammar.Rule], file: Optional[IO[Text]]):
         super().__init__(rules, file)
         self.callmakervisitor = CCallMakerVisitor(self)
+        self._varname_counter = itertools.count()
+
+    def unique_varname(self, name="tmpvar"):
+        return name + "_" + str(next(self._varname_counter))
+
+    def call_with_errorcheck(self, call_text, returnval="", goto_target=None):
+        error_var = self.unique_varname()
+        self.print(f"int {error_var} = {call_text}")
+        self.print(f"if ({error_var}) {{")
+        with self.indent():
+            if goto_target:
+                self.print(f"goto {goto_target};")
+            else:
+                self.print(f"return {returnval};")
+        self.print(f"}}")
+
+    def out_of_memory(self, expr, text="Parser out of memory", returnval="", goto_target=None):
+        self.print(f"if ({expr}) {{")
+        with self.indent():
+            self.print(f'PyErr_Format(PyExc_MemoryError, "{text}");')
```

Comment:
Maybe I'm prematurely optimizing, but I wonder if this warrants a tiny C-level helper that you can call without arguments that raises MemoryError with a fixed error string. Exactly where it ran out of memory is unlikely to be of interest (and if you really care you can use gdb/lldb).

Reply:
You mean a function that just calls `PyErr_Format(PyExc_MemoryError, "The parser ran out of memory");` or a  macro that also returns?

Reply:
I meant a function. It's premature because I'm trying to reduce binary code size in a prototype. (I do expect that there will be a lot of generated code. But still I'm being premature.)